python/mxnet/module/base_module.py

```diff
@@ -78,6 +79,17 @@ def _parse_data_desc(data_names, label_names, data_shapes, label_shapes):
     return data_shapes, label_shapes
 
 
+def _check_data_names(eval_data, data_names):
+    """ Check if iterator data names match the data names provided in module"""
+    if isinstance(eval_data, io.NDArrayIter) and \
```

Comment:
This is still special casing NDArrayIter. The same problem would happen for all dataiters if names are incorrect.
Why do we need renamed_data attribute?

Reply:
In NDArrayIter, user can provide data in three formats:
data (array or list of array or dict of string to array) 

if the data is not a dict, mxnet will create a dict with built-in data_names, these would not match the data names provided by the user . We need to only match in the scenario if the a dict value was provided by the user while creating NDArrayIter